### PR TITLE
Add support for Aion [chain integration]

### PR DESCRIPTION
**Pull request by ivan-trust**
## Description

Added support for Aion.

Updated HDWallet, PrivateKey and PublicKey to support Aion key format.

## Checklist

- [x] Add tests to cover changes as needed.
- [x] Return correct curve and purpose in `src/Coin.cpp`.
- [x] Derivation from private key in `src/Coin.cpp`.
- [x] Validation in `src/Coin.cpp`.

### CODE REVIEW

File: src/Aion/Transaction.cpp [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+using namespace TW::Aion;
+using boost::multiprecision::uint128_t;
+
+Data Transaction::encode() const noexcept {
+    auto encoded = Data();
+    append(encoded, Ethereum::RLP::encode(nonce));
+    append(encoded, Ethereum::RLP::encode(to.bytes));
+    append(encoded, Ethereum::RLP::encode(amount));
+    append(encoded, Ethereum::RLP::encode(payload));
+    append(encoded, Ethereum::RLP::encode(uint128_t(155157377101))); // Huge timestamp
```
Comment (hewigovens) [inline src/Aion/Transaction.cpp]: what's this hardcoded value?

File: src/HDWallet.cpp [hunk 2]
```diff
@@ -146,22 +147,46 @@ std::optional<std::string> HDWallet::getAddressFromExtended(const std::string& e
...
+        
         hdnode_private_ckd(&node, purpose | 0x80000000);
         hdnode_private_ckd(&node, coin | 0x80000000);
         hdnode_private_ckd(&node, account | 0x80000000);
-        hdnode_private_ckd(&node, change);
-        hdnode_private_ckd(&node, address);
+        
+        auto correctedChange = change;
+        auto correctedAddress = address;
+        if (coin == TWCoinTypeAion) {
```
Comment (hewigovens) [inline src/HDWallet.cpp]: do `switch` here as well?

**Comment by hewigovens** [comment]
could you please rebase master and fix CI (Failed - Wallet Core) ? it would be nice to have swift interface tests (test AionAddress.swift and AionSigner.swift)

File: src/HDWallet.cpp (hunk 2 above)
Comment (vikmeup) [reply src/HDWallet.cpp]: I would find a way to not have a switch for coins in this function, @alejandro-isaza might have more insight on this

Comment (hewigovens) [reply src/HDWallet.cpp]: we may not need change / address for curve ed25519 this? https://github.com/satoshilabs/slips/blob/master/slip-0010.md

File: include/TrustWalletCore/TWPublicKey.h [hunk 3]
```diff
@@ -38,6 +41,9 @@ TWData *_Nonnull TWPublicKeyData(struct TWPublicKey pk);
 TW_EXPORT_METHOD
 bool TWPublicKeyVerify(struct TWPublicKey pk, TWData *_Nonnull signature, TWData *_Nonnull message);
 
+TW_EXPORT_METHOD
+bool TWPublicKeyVerifyEd25519(struct TWPublicKey pk, TWData *_Nonnull signature, TWData *_Nonnull message);
```
Comment (terorie) [inline include/TrustWalletCore/TWPublicKey.h]: https://github.com/TrustWallet/wallet-core/pull/76
would expose a nicer C and C++ API

File: src/HDWallet.cpp (hunk 2 above)
Comment (ivan-trust) [reply src/HDWallet.cpp]: I did `switch` for now.

File: src/Aion/Transaction.cpp (hunk 1 above)
Comment (ivan-trust) [reply src/Aion/Transaction.cpp]: Aion-web3 takes [current timestamp * 1000](https://github.com/aionnetwork/aion_web3/blob/52453c82aa353f74f66f17f67d7d1f796c238d99/packages/web3-eth-accounts/src/index.js#L211) when building a transaction, if it's not provided by user. I just hardcoded some huge timestamp.

**Comment by ivan-trust** [comment]
> could you please rebase master and fix CI (Failed - Wallet Core) ? it would be nice to have swift interface tests (test AionAddress.swift and AionSigner.swift)

Rebased master, added Swift tests.

File: include/TrustWalletCore/TWPublicKey.h (hunk 3 above)
Comment (ivan-trust) [reply include/TrustWalletCore/TWPublicKey.h]: Great, I'll update my PR.

Comment (terorie) [reply include/TrustWalletCore/TWPublicKey.h]: @ivan-trust Please don't, they were just reverted 😬 
@alejandro-isaza Is working on a different API, which will be released soon.

File: swift/Sources/Blockchains/Aion.swift [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    public override func address(data: Data) -> Address? {
+        return AionAddress(data: data)
+    }
+    
+}
+
+extension AionAddress: Address {
+    
+    public static func isValid(data: Data) -> Bool {
+        if let _ = AionAddress(data: data) {
```
Comment (vikmeup) [inline swift/Sources/Blockchains/Aion.swift]: `return AionAddress(data: data) != nil`

**Comment by hewigovens** [comment]
overall looks good, Is there any other HD wallets support Aion? so we can cross verify the addresses derived from phrases.

**Comment by alejandro-isaza** [comment]
Updated checklist, please complete.

**Comment by ivan-trust** [comment]
> Updated checklist, please complete.
> https://github.com/TrustWallet/wallet-core/blob/master/.github/ISSUE_TEMPLATE/new_blockchain.md#checklist

✅

**Pull request merged**